machine/v1/types_alibabaprovider.go

```diff
@@ -0,0 +1,340 @@
+
+	// SecurityGroups is a list of security group references to assign to the instance. A reference holds either the security group ID
+	// or the required tags to search. The  limit of N values varies based on the maximum number of security groups to which an instance can belong.
+	// For more information, see the "Security group limits" section in Limits.
+	// https://www.alibabacloud.com/help/doc-detail/101348.htm?spm=a2c63.p38356.879954.48.78f0199aX3dfIE
+	SecurityGroups []AlibabaResourceReference `json:"securityGroups,omitempty"`
+
+	// Bandwidth describes the internet bandwidth strategy for the instance
+	// +optional
+	Bandwidth BandwidthProperties `json:"bandwidth,omitempty"`
+
+	// SystemDisk holds the properties regarding the system disk for the instance
+	// +optional
+	SystemDisk SystemDiskProperties `json:"systemDisk,omitempty"`
+
+	// VSwitch is a reference to the vswitch to use for this instance
+	// This parameter is required when you create an instance of the VPC type.
+	// You can call the DescribeVSwitches operation to query the created vSwitches.
+	VSwitch AlibabaResourceReference `json:"vSwitch,omitempty"`
+
+	// RAMRoleName is the name of the instance Resource Access Management (RAM) role. This allows the instance to perform API calls as this specified RAM role.
+	// +optional
+	RAMRoleName string `json:"ramRoleName,omitempty"`
+
+	// ResourceGroupID is the unique ID of the resource group to which to assign the instance.
+	// +optional
+	ResourceGroupID string `json:"resourceGroupId,omitempty"`
+
+	// Tenancy specifies whether to create the instance on a dedicated host.
+	// Valid values:
+	//
+	// default: creates the instance on a non-dedicated host.
+	// host: creates the instance on a dedicated host. If you do not specify the DedicatedHostID parameter, Alibaba Cloud automatically selects a dedicated host for the instance.
+	// Empty value means no opinion and the platform chooses the a default, which is subject to change over time.
+	// Currently the default is `default`.
+	// +optional
+	Tenancy InstanceTenancy `json:"tenancy,omitempty"`
+
+	// UserDataSecret contains a local reference to a secret that contains the
+	// UserData to apply to the instance
+	// +optional
+	UserDataSecret *corev1.LocalObjectReference `json:"userDataSecret,omitempty"`
+
+	// CredentialsSecret is a reference to the secret with alibabacloud credentials. Otherwise, defaults to permissions
+	// provided by attached RAM role where the actuator is running.
+	// +optional
+	CredentialsSecret *corev1.LocalObjectReference `json:"credentialsSecret,omitempty"`
+
+	// Tags are the set of metadata to add to an instance.
+	// +optional
+	Tags []Tag `json:"tag,omitempty"`
+}
+
+// ResourceTagReference is a reference to a specific AlibabaCloud resource by ID, or tags.
+// Only one of ID or Tags may be specified. Specifying more than one will result in
+// a validation error.
+type AlibabaResourceReference struct {
+	// ID of resource
+	// +optional
```

Comment:
this is used in spec, how can ID be empty in spec?

Reply:
This is similar to the AWS provider field. This can either hold an ID to an AlibabaResourceReference or this struct holds the necessary tags that can be used to search for the resource. (e.g. securityGroups. Note, no ID but tags required to find the resource ID).
```
            securityGroups:
            - tags:
              - Key: kubernetes.io/cluster/test-2qzw5
                Value: owned
              - Key: GISV
                Value: ocp
              - Key: sigs.k8s.io/cloud-provider-alibaba/origin
                Value: ocp
              - Key: Name
                Value: test-2qzw5-sg-worker

```